// @astrojs/parser/dist/parse/read/script.js
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// src/parse/read/script.ts
__markAsModule(exports);
__export(exports, {
  default: () => read_script
});
var script_closing_tag = "</script>";
function get_context(parser, attributes, start) {
  const context = attributes.find((attribute) => attribute.name === "astro");
  if (!context)
    return "runtime";
  if (context.value === true)
    return "setup";
  if (context.value.length !== 1 || context.value[0].type !== "Text") {
    parser.error({
      code: "invalid-script",
      message: "astro attribute must be static"
    }, start);
  }
  const value = context.value[0].data;
  if (value !== "setup") {
    parser.error({
      code: "invalid-script",
      message: 'If the "astro" attribute has a value, its value must be "setup"'
    }, context.start);
  }
  return value;
}
function read_script(parser, start, attributes) {
  const script_start = parser.index;
  const script_end = parser.template.indexOf(script_closing_tag, script_start);
  if (script_end === -1) {
    parser.error({
      code: "unclosed-script",
      message: "<script> must have a closing tag"
    });
  }
  const source = parser.template.slice(0, script_start).replace(/[^\n]/g, " ") + parser.template.slice(script_start, script_end);
  parser.index = script_end + script_closing_tag.length;
  return {
    type: "Script",
    start,
    end: parser.index,
    context: get_context(parser, attributes, start),
    content: source
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {});
