// @astrojs/parser/dist/utils/namespaces.js
var __defProp = Object.defineProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};

// src/utils/namespaces.ts
__markAsModule(exports);
__export(exports, {
  foreign: () => foreign,
  html: () => html,
  mathml: () => mathml,
  namespaces: () => namespaces,
  svg: () => svg,
  valid_namespaces: () => valid_namespaces,
  xlink: () => xlink,
  xml: () => xml,
  xmlns: () => xmlns
});
var foreign = "https://svelte.dev/docs#svelte_options";
var html = "http://www.w3.org/1999/xhtml";
var mathml = "http://www.w3.org/1998/Math/MathML";
var svg = "http://www.w3.org/2000/svg";
var xlink = "http://www.w3.org/1999/xlink";
var xml = "http://www.w3.org/XML/1998/namespace";
var xmlns = "http://www.w3.org/2000/xmlns";
var valid_namespaces = ["foreign", "html", "mathml", "svg", "xlink", "xml", "xmlns", foreign, html, mathml, svg, xlink, xml, xmlns];
var namespaces = {foreign, html, mathml, svg, xlink, xml, xmlns};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  foreign,
  html,
  mathml,
  namespaces,
  svg,
  valid_namespaces,
  xlink,
  xml,
  xmlns
});
