// @astrojs/parser/dist/interfaces.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __commonJS = (cb, mod) => function __require() {
  return mod || (0, cb[Object.keys(cb)[0]])((mod = {exports: {}}).exports, mod), mod.exports;
};
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// ../../node_modules/locate-character/dist/locate-character.umd.js
var require_locate_character_umd = __commonJS({
  "../../node_modules/locate-character/dist/locate-character.umd.js"(exports, module2) {
    (function(global, factory) {
      typeof exports === "object" && typeof module2 !== "undefined" ? factory(exports) : typeof define === "function" && define.amd ? define(["exports"], factory) : factory(global.locateCharacter = global.locateCharacter || {});
    })(exports, function(exports2) {
      "use strict";
      function getLocator(source, options) {
        if (options === void 0) {
          options = {};
        }
        var offsetLine = options.offsetLine || 0;
        var offsetColumn = options.offsetColumn || 0;
        var originalLines = source.split("\n");
        var start = 0;
        var lineRanges = originalLines.map(function(line, i2) {
          var end = start + line.length + 1;
          var range = {start, end, line: i2};
          start = end;
          return range;
        });
        var i = 0;
        function rangeContains(range, index) {
          return range.start <= index && index < range.end;
        }
        function getLocation(range, index) {
          return {line: offsetLine + range.line, column: offsetColumn + index - range.start, character: index};
        }
        function locate3(search, startIndex) {
          if (typeof search === "string") {
            search = source.indexOf(search, startIndex || 0);
          }
          var range = lineRanges[i];
          var d = search >= range.end ? 1 : -1;
          while (range) {
            if (rangeContains(range, search))
              return getLocation(range, search);
            i += d;
            range = lineRanges[i];
          }
        }
        ;
        return locate3;
      }
      function locate2(source, search, options) {
        if (typeof options === "number") {
          throw new Error("locate takes a { startIndex, offsetLine, offsetColumn } object as the third argument");
        }
        return getLocator(source, options)(search, options && options.startIndex);
      }
      exports2.getLocator = getLocator;
      exports2.locate = locate2;
      Object.defineProperty(exports2, "__esModule", {value: true});
    });
  }
});

// src/interfaces.ts
__markAsModule(exports);
__export(exports, {
  CompileError: () => CompileError
});

// src/utils/error.ts
var import_locate_character = __toModule(require_locate_character_umd());

// src/utils/get_code_frame.ts
function tabs_to_spaces(str) {
  return str.replace(/^\t+/, (match) => match.split("	").join("  "));
}
function get_code_frame(source, line, column) {
  const lines = source.split("\n");
  const frame_start = Math.max(0, line - 2);
  const frame_end = Math.min(line + 3, lines.length);
  const digits = String(frame_end + 1).length;
  return lines.slice(frame_start, frame_end).map((str, i) => {
    const isErrorLine = frame_start + i === line;
    const line_num = String(i + frame_start + 1).padStart(digits, " ");
    if (isErrorLine) {
      const indicator = " ".repeat(digits + 2 + tabs_to_spaces(str.slice(0, column)).length) + "^";
      return `${line_num}: ${tabs_to_spaces(str)}
${indicator}`;
    }
    return `${line_num}: ${tabs_to_spaces(str)}`;
  }).join("\n");
}

// src/utils/error.ts
var CompileError = class extends Error {
  constructor({code, filename, start, end, message}) {
    super(message);
    this.start = (0, import_locate_character.locate)(code, start, {offsetLine: 1});
    this.end = (0, import_locate_character.locate)(code, end || start, {offsetLine: 1});
    this.filename = filename;
    this.message = message;
    this.frame = get_code_frame(code, this.start.line - 1, this.start.column);
  }
  toString() {
    return `${this.filename}:${this.start.line}:${this.start.column}
	${this.message}
${this.frame}`;
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  CompileError
});
